pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,170 @@
+type IndexContext struct {
+	Index uint64
+	Ctx   context.Context
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []IndexContext {
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// Close will unregister all the traces and log them. It should be called if the
+// replica is destroyed as it is possible some traces were never closed.
+func (r *RaftTracer) Close() {
+	r.m.Range(func(index uint64, val *traceValue) bool {
+		log.VEventfDepth(val.ctx, 5, 3, "cleanup log mark (i%d) during close", index)
+		trace := val.cleanup().String()
+		// Consider writing this to a different log if it becomes too noisy.
+		log.InfofDepth(context.Background(), 5, "%s", trace)
+		return true
+	})
+}
+
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []IndexContext {
+	var savedCtxs []IndexContext
+	r.m.Range(func(index uint64, val *traceValue) bool {
+		switch m.Type {
+		case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+			if inRange(index, m.Entries) {
+				savedCtxs = append(savedCtxs, IndexContext{Index: index, Ctx: val.ctx})
+				log.VEventfDepth(val.ctx, 4, 3, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
+				if val.baseCtx != nil {
+					log.VEventfDepth(val.baseCtx.Context(), 4, 3, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
+				}
+			}
+		}
+		return true
+	})
+	return savedCtxs
+}
+
+func (r *RaftTracer) traceIfPast(m raftpb.Message) {
+	if m.Reject {
+		return
+	}
+	r.m.Range(func(index uint64, val *traceValue) bool {
+		shouldLog := false
+		shouldUnregister := false
+		switch m.Type {
+		case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp:
+			shouldLog = m.Index >= index
+		case raftpb.MsgStorageApplyResp:
+			var msgIndex uint64
+			if len(m.Entries) != 0 {
+				index = m.Entries[len(m.Entries)-1].Index
+			}
+			shouldLog = index > msgIndex
+			if shouldLog {
+				shouldUnregister = true
+			}
+		}
+		if shouldLog {
+			log.VEventfDepth(val.ctx, 4, 3, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
+			if val.baseCtx != nil {
+				log.VEventfDepth(val.baseCtx.Context(), 4, 3, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
+			}
+		}
+		if shouldUnregister {
+			// We unregister the index here because we are now "done" with
+			// this and don't expect more useful entries. There could still
+			// be non-quorum messages, but they don't impact anything the
+			// client cares about.
+			r.unregisterIndex(index)
```

Comment:
This only unregisters one `index`, but we need to unregister the entire prefix up to this index.

`MsgStorageApplyResp` is not scoped to a particular entry index. It communicates that all entries up to the `LogMark` in it are synced.

Reply:
This is inside the Range loop above, so I think it will unregister all, but let me double check.

Reply:
I wonder if we should put a cap on "max outstanding traces". It would ideally never be hit in practice, but by setting to something like 20 we avoid any of the risk of runaway proposals. Thoughts?